habitat_baselines/common/rollout_storage.py

```diff
@@ -4,237 +4,196 @@
+        if num_simulators % num_mini_batch != 0:
+            warnings.warn(
+                "Number of simulators ({}) is not a multiple of the number of mini batches ({})".format(
+                    num_simulators, num_mini_batch
                 )
-
-                adv_targ.append(advantages[: self.step, ind])
-
-            T, N = self.step, num_envs_per_batch
-
-            # These are all tensors of size (T, N, -1)
-            for sensor in observations_batch:
-                observations_batch[sensor] = torch.stack(
-                    observations_batch[sensor], 1
-                )
-
-            actions_batch = torch.stack(actions_batch, 1)
-            prev_actions_batch = torch.stack(prev_actions_batch, 1)
-            value_preds_batch = torch.stack(value_preds_batch, 1)
-            return_batch = torch.stack(return_batch, 1)
-            masks_batch = torch.stack(masks_batch, 1)
-            old_action_log_probs_batch = torch.stack(
-                old_action_log_probs_batch, 1
             )
-            adv_targ = torch.stack(adv_targ, 1)
-
-            # States is just a (num_recurrent_layers, N, -1) tensor
-            recurrent_hidden_states_batch = torch.stack(
-                recurrent_hidden_states_batch, 1
-            )
-
-            # Flatten the (T, N, ...) tensors to (T * N, ...)
-            for sensor in observations_batch:
-                observations_batch[sensor] = self._flatten_helper(
-                    T, N, observations_batch[sensor]
-                )
-
-            actions_batch = self._flatten_helper(T, N, actions_batch)
-            prev_actions_batch = self._flatten_helper(T, N, prev_actions_batch)
-            value_preds_batch = self._flatten_helper(T, N, value_preds_batch)
-            return_batch = self._flatten_helper(T, N, return_batch)
-            masks_batch = self._flatten_helper(T, N, masks_batch)
-            old_action_log_probs_batch = self._flatten_helper(
-                T, N, old_action_log_probs_batch
-            )
-            adv_targ = self._flatten_helper(T, N, adv_targ)
-
-            yield (
-                observations_batch,
-                recurrent_hidden_states_batch,
-                actions_batch,
-                prev_actions_batch,
-                value_preds_batch,
-                return_batch,
-                masks_batch,
-                old_action_log_probs_batch,
-                adv_targ,
-            )
-
-    @staticmethod
-    def _flatten_helper(t: int, n: int, tensor: torch.Tensor) -> torch.Tensor:
-        r"""Given a tensor of size (t, n, ..), flatten it to size (t*n, ...).
-
-        Args:
-            t: first dimension of tensor.
-            n: second dimension of tensor.
-            tensor: target tensor to be flattened.
-
-        Returns:
-            flattened tensor of size (t*n, ...)
-        """
-        return tensor.view(t * n, *tensor.size()[2:])
+        for inds in torch.randperm(num_simulators).chunk(num_mini_batch):
```

Comment:
By the way, why do we want random permutation here?

Reply:
That way the mini batches are different for each epoch

Reply:
That means, if we have `num_simulators > num_mini_batch` the way the ids of environments/episodes will change how it's presented in mini_batches.

Reply:
Yep.  Just to clarify, we have always done this, this is just a cleaner and more robust way to do it.